File: src/main/java/net/ddns/rarnold/randomworkoutroutine/service/ItemService.java
package net.ddns.rarnold.randomworkoutroutine.service;

import lombok.RequiredArgsConstructor;
import net.ddns.rarnold.randomworkoutroutine.model.Filter;
import net.ddns.rarnold.randomworkoutroutine.model.entity.Item;
import net.ddns.rarnold.randomworkoutroutine.repository.ItemRepository;

import java.util.*;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public abstract class ItemService<T extends Item> {

    protected final ItemRepository<T> repository;

    public Optional<T> getById(UUID id) {
        return repository.findById(id);
    }

    // TODO: Don't allow saving of routines and groups with no children. This is a problem when the
    //  user deletes a group or exercise that is the only child of a routine or group. Maybe just
    //  stop the user starting sessions if the routine or a group in the routine has no children.
    public void save(T item) {
        repository.save(item);
    }

    public void delete(T item) {
        repository.delete(item);
    }

    public List<T> getNames() {
        return transformIdAndNameToItem(repository.findAllIdsAndNames());
    }

    public List<T> search(Filter filter) {
        String searchTerm = "";
        Set<String> excludedNames = Set.of("");
        if (filter != null) {
            if (filter.getSearchTerm() != null) {
                searchTerm = filter.getSearchTerm();
            }
            if (filter.getExcludedTerms() != null && !filter.getExcludedTerms().isEmpty()) {
                excludedNames = filter.getExcludedTerms();
            }
        }
        return transformIdAndNameToItem(repository.findAllIdsAndNames(searchTerm, excludedNames));
    }

    protected List<T> transformIdAndNameToItem(List<Object[]> idsAndNames) {
        return idsAndNames.stream()
                .map(objects -> createItem((UUID) objects[0], (String) objects[1]))
                .collect(Collectors.toList());
    }

    protected abstract T createItem(UUID id, String name);
}

File: web/src/main/kotlin/uk/co/rafearnold/randomworkoutroutine/web/service/ItemService.kt
package uk.co.rafearnold.randomworkoutroutine.web.service

import uk.co.rafearnold.randomworkoutroutine.model.SimpleItem
import uk.co.rafearnold.randomworkoutroutine.model.SimpleItemImpl
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Item
import uk.co.rafearnold.randomworkoutroutine.web.repository.ItemRepository
import java.util.*

/**
 * Abstract service for retrieving and manipulating [Item] entities.
 */
abstract class ItemService<T : Item>(protected val repository: ItemRepository<T>) {

    fun getById(id: UUID): Optional<T> = repository.findById(id)

    fun save(item: T) {
        repository.save(item)
    }

    open fun delete(item: T) {
        repository.delete(item)
    }

    /**
     * Retrieves all saved [T] items and returns them as [SimpleItem] objects.
     */
    fun getNames(): List<SimpleItem> = repository.findAllIdsAndNames()

    /**
     * Retrieves all saved [T] items that match [searchTerm], excluding any that match any items in
     * [excludedNames], and returns them as [SimpleItem] objects.
     */
    fun search(searchTerm: String, excludedNames: Set<String>): List<SimpleItem> =
        repository.findAllIdsAndNames(searchTerm, if (excludedNames.isEmpty()) setOf("") else excludedNames)

    /**
     * Transforms each object array in [idsAndNames] into a [SimpleItem]. 
     */
    private fun transformIdAndNameToItem(idsAndNames: List<Array<Any>>): List<SimpleItem> =
        idsAndNames.map { SimpleItemImpl(it[0] as UUID, it[1] as String) }
}
